crates/sol-macro-gen/src/sol_macro_gen.rs

```diff
@@ -0,0 +1,367 @@
+    }
+
+    pub fn populate_expansion(&mut self, bindings_path: &Path) -> Result<()> {
+        for instance in &mut self.instances {
+            let path = bindings_path.join(format!("{}.rs", instance.name.to_lowercase()));
+            let expansion =
+                fs::read_to_string(path).map_err(|e| eyre::eyre!("Failed to read file: {e}"))?;
+
+            let tokens = TokenStream::from_str(&expansion)
+                .map_err(|e| eyre::eyre!("Failed to parse TokenStream: {e}"))?;
+            instance.expansion = Some(tokens);
+        }
+        Ok(())
+    }
+
+    pub fn generate_bindings(&mut self) -> Result<()> {
+        for instance in &mut self.instances {
+            let (mut json_abi, maybe_bytecode) = instance.get_json_abi()?;
+
+            json_abi.dedup();
+            let sol_str = json_abi.to_sol(&instance.name, None);
+
+            let ident_name: Ident = Ident::new(&instance.name, Span::call_site());
+
+            let tokens = tokens_for_sol(&ident_name, &sol_str)
+                .map_err(|e| eyre::eyre!("Failed to get sol tokens: {e}"))?;
+
+            let tokens = if let Some(bytecode) = maybe_bytecode {
+                let bytecode = proc_macro2::TokenStream::from_str(&bytecode).map_err(|e| {
+                    eyre::eyre!("Failed to convert bytecode String to TokenStream {e}")
+                })?;
+                quote::quote! {
+                    #[derive(Debug)]
+                    #[sol(rpc, bytecode = #bytecode)]
+                    #tokens
+                }
+            } else {
+                quote::quote! {
+                    #[derive(Debug)]
+                    #[sol(rpc)]
+                    #tokens
+                }
+            };
+
+            let input: SolInput =
+                syn::parse2(tokens).map_err(|e| eyre::eyre!("Failed to parse SolInput: {e}"))?;
+
+            let SolInput { attrs: _attrs, path: _path, kind } = input;
+
+            let tokens = match kind {
+                SolInputKind::Sol(file) => {
+                    expand(file).map_err(|e| eyre::eyre!("Failed to expand SolInput: {e}"))?
+                }
+                _ => unreachable!(),
+            };
+
+            instance.expansion = Some(tokens);
+        }
+
+        Ok(())
+    }
+
+    pub fn write_to_crate(
+        &mut self,
+        name: &str,
+        version: &str,
+        bindings_path: &Path,
+        single_file: bool,
+        alloy_version: String,
+    ) -> Result<()> {
+        self.generate_bindings()?;
+
+        let src = bindings_path.join("src");
+
+        let _ = fs::create_dir_all(&src);
+
+        // Write Cargo.toml
+        let cargo_toml_path = bindings_path.join("Cargo.toml");
+        let toml_contents = format!(
+            r#"[package]
+name = "{}"
+version = "{}"
+edition = "2021"
+
+[dependencies]
+alloy-sol-types = "{}"
+alloy-contract = {{ git = "https://github.com/alloy-rs/alloy", rev = "64feb9b" }}"#,
```

Comment:
this is bad, should not hardcode rev
```suggestion
alloy-contract = {{ git = "https://github.com/alloy-rs/alloy" }}"#,
```